### PR TITLE
Tweak ContinuableScopeManager for better inlining

### PR DESCRIPTION
**Pull request by richardstartin**
After #1658 there was a strange increase in CPU utilisation in the reliability environment, which seems to have been caused by crossing an inlining threshold. This PR makes several tweaks which generally move exceptional code (e.g. logging or handling unexpected cases) out of hot methods, and reduces the number of accesses to the `ThreadLocal`. Here's the inlining in spring-petclinic

Before #1658 `58cf75de66a7f60a0966411e4812c1228af87a7d `

```
                              @ 4   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::active (10 bytes)
                              @ 29   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activeSpan (25 bytes)
                              @ 21   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::access$000 (4 bytes)
                              @ 5   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activate (75 bytes)   callee is too large
                              @ 26   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::incrementReferences (10 bytes)
                              @ 39   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::depth (5 bytes)
                              @ 71   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::handleSpan (7 bytes)
                                @ 3   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::handleSpan (33 bytes)   callee is too large
                              @ 3   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::handleSpan (33 bytes)
                                @ 16   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::<init> (100 bytes)   callee is too large
                              @ 16   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::<init> (100 bytes)   callee is too large
                              @ 91   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::depth (5 bytes)
                              @ 29   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activeSpan (25 bytes)   inline (hot)
                               \-> TypeProfile (19703/19703 counts) = datadog/trace/agent/core/scopemanager/ContinuableScopeManager
                              @ 5   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activate (75 bytes)   inline (hot)
                               \-> TypeProfile (26941/26941 counts) = datadog/trace/agent/core/scopemanager/ContinuableScopeManager
                                @ 39   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::depth (5 bytes)   accessor
                                @ 71   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::handleSpan (7 bytes)   inline (hot)
                                  @ 3   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::handleSpan (33 bytes)   inline (hot)
                                    @ 16   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::<init> (100 bytes)   inline (hot)
                                      @ 91   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::depth (5 bytes)   accessor
                              @ 39   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::depth (5 bytes)   accessor
                              @ 71   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::handleSpan (7 bytes)   inline (hot)
                                @ 3   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::handleSpan (33 bytes)   inline (hot)
                                  @ 16   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::<init> (100 bytes)   inline (hot)
                                    @ 91   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::depth (5 bytes)   accessor
                              @ 16   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::<init> (100 bytes)   inline (hot)
                                @ 91   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::depth (5 bytes)   accessor
                                 \-> TypeProfile (7099/7099 counts) = datadog/trace/agent/core/scopemanager/ContinuableScopeManager$ContinuableScope
                                @ 166   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::close (85 bytes)   inline (hot)
                              @ 4   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::active (10 bytes)
                              @ 4   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activeSpan (25 bytes)
```

master `fdc5b576f3413cf4e66f6d2acf0dd71308612b3e `

```
                              @ 4   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::active (10 bytes)
                              @ 29   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activeSpan (25 bytes)
                              @ 6   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activate (87 bytes)   callee is too large
                              @ 26   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::incrementReferences (10 bytes)
                              @ 39   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::depth (5 bytes)
                              @ 83   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::handleSpan (38 bytes)   callee is too large
                              @ 17   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::<init> (106 bytes)   callee is too large
                              @ 91   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::depth (5 bytes)
                              @ 21   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::access$000 (4 bytes)
                              @ 42   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::access$100 (5 bytes)
                              @ 70   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::access$100 (5 bytes)
                              @ 88   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::access$200 (5 bytes)
                              @ 29   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activeSpan (25 bytes)   inline (hot)
                               \-> TypeProfile (13319/13319 counts) = datadog/trace/agent/core/scopemanager/ContinuableScopeManager
                              @ 6   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activate (87 bytes)   inline (hot)
                               \-> TypeProfile (23562/23562 counts) = datadog/trace/agent/core/scopemanager/ContinuableScopeManager
                                @ 39   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::depth (5 bytes)   accessor
                                @ 83   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::handleSpan (38 bytes)   inline (hot)
                                  @ 17   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::<init> (106 bytes)   inline (hot)
                                    @ 91   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::depth (5 bytes)   accessor
                                                 \-> TypeProfile (28082/28082 counts) = datadog/trace/agent/core/scopemanager/ContinuableScopeManager$ContinuableScope
                                                @ 166   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::close (151 bytes)   inline (hot)
                               \-> TypeProfile (66383/66383 counts) = datadog/trace/agent/core/scopemanager/ContinuableScopeManager$ContinuableScope
                              @ 166   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::close (151 bytes)   inline (hot)
                              @ 4   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activeSpan (25 bytes)
```

Here, `2ea761e86fd5e4eb0d072be0e725af08737b0bca `, everything gets inlined except for the exceptional methods.

```
                              @ 4   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::active (10 bytes)
                              @ 29   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activeSpan (25 bytes)
                              @ 6   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activate (38 bytes)   callee is too large
                              @ 26   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::incrementReferences (10 bytes)
                              @ 34   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activate (38 bytes)   callee is too large
                              @ 9   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::access$000 (5 bytes)
                              @ 25   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::depthLimitExceeded (18 bytes)
                              @ 34   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::handleSpan (73 bytes)   callee is too large
                              @ 23   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::handleSpan (73 bytes)   callee is too large
                              @ 52   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::<init> (60 bytes)   callee is too large
                              @ 47   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::childDepth (16 bytes)
                                @ 10   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::access$000 (5 bytes)
                              @ 10   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::access$000 (5 bytes)
                              @ 22   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::closeExceptionally (84 bytes)   callee is too large
                              @ 29   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activeSpan (25 bytes)   inline (hot)
                               \-> TypeProfile (21384/21384 counts) = datadog/trace/agent/core/scopemanager/ContinuableScopeManager
                              @ 6   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activate (38 bytes)   inline (hot)
                               \-> TypeProfile (34812/34812 counts) = datadog/trace/agent/core/scopemanager/ContinuableScopeManager
                                @ 34   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activate (38 bytes)   inline (hot)
                                  @ 9   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::access$000 (5 bytes)   accessor
                                  @ 34   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::handleSpan (73 bytes)   inline (hot)
                                    @ 52   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::<init> (60 bytes)   inline (hot)
                                      @ 47   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::childDepth (16 bytes)   inline (hot)
                                        @ 10   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::access$000 (5 bytes)   accessor
                              @ 34   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::activate (38 bytes)   inline (hot)
                                @ 9   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::access$000 (5 bytes)   accessor
                                @ 34   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::handleSpan (73 bytes)   inline (hot)
                                  @ 52   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::<init> (60 bytes)   inline (hot)
                                    @ 47   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::childDepth (16 bytes)   inline (hot)
                                      @ 10   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::access$000 (5 bytes)   accessor
                              @ 9   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::access$000 (5 bytes)   accessor
                              @ 34   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::handleSpan (73 bytes)   inline (hot)
                                @ 52   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::<init> (60 bytes)   inline (hot)
                                  @ 47   datadog.trace.agent.core.scopemanager.ContinuableScopeManager::childDepth (16 bytes)   inline (hot)
                                    @ 10   datadog.trace.agent.core.scopemanager.ContinuableScopeManager$ContinuableScope::access$000 (5 bytes)   accessor
```

It's difficult to provoke the behaviour here (and it will vary from application to application too) in a microbenchmark, because it depends more on the context of the caller than it does on the local shape of the code. Nevertheless, I added a microbenchmark and small gains can be seen from the reduction in TLS lookups, but there isn't enough code to stress inlining heuristics.

master `fdc5b576f3413cf4e66f6d2acf0dd71308612b3e`

```
Benchmark                                  (depth)  (maxDepth)  (strict)    (work)  Mode  Cnt    Score   Error  Units
ContinuableScopeManagerBenchmark.activate        1           0      true     LIGHT  avgt    5   59.654 ± 1.406  ns/op
ContinuableScopeManagerBenchmark.activate        1           0      true  MODERATE  avgt    5   63.831 ± 1.280  ns/op
ContinuableScopeManagerBenchmark.activate        1           0     false     LIGHT  avgt    5   58.787 ± 0.713  ns/op
ContinuableScopeManagerBenchmark.activate        1           0     false  MODERATE  avgt    5   64.287 ± 0.538  ns/op
ContinuableScopeManagerBenchmark.activate        1         100      true     LIGHT  avgt    5   59.812 ± 0.646  ns/op
ContinuableScopeManagerBenchmark.activate        1         100      true  MODERATE  avgt    5   63.968 ± 0.715  ns/op
ContinuableScopeManagerBenchmark.activate        1         100     false     LIGHT  avgt    5   59.739 ± 0.654  ns/op
ContinuableScopeManagerBenchmark.activate        1         100     false  MODERATE  avgt    5   64.642 ± 0.841  ns/op
ContinuableScopeManagerBenchmark.activate        3           0      true     LIGHT  avgt    5  100.233 ± 1.734  ns/op
ContinuableScopeManagerBenchmark.activate        3           0      true  MODERATE  avgt    5  122.567 ± 1.019  ns/op
ContinuableScopeManagerBenchmark.activate        3           0     false     LIGHT  avgt    5   98.074 ± 0.498  ns/op
ContinuableScopeManagerBenchmark.activate        3           0     false  MODERATE  avgt    5  119.912 ± 0.695  ns/op
ContinuableScopeManagerBenchmark.activate        3         100      true     LIGHT  avgt    5   97.573 ± 1.271  ns/op
ContinuableScopeManagerBenchmark.activate        3         100      true  MODERATE  avgt    5  119.201 ± 0.426  ns/op
ContinuableScopeManagerBenchmark.activate        3         100     false     LIGHT  avgt    5   96.639 ± 1.217  ns/op
ContinuableScopeManagerBenchmark.activate        3         100     false  MODERATE  avgt    5  123.188 ± 1.117  ns/op
ContinuableScopeManagerBenchmark.activate        5           0      true     LIGHT  avgt    5  138.134 ± 2.360  ns/op
ContinuableScopeManagerBenchmark.activate        5           0      true  MODERATE  avgt    5  171.540 ± 2.626  ns/op
ContinuableScopeManagerBenchmark.activate        5           0     false     LIGHT  avgt    5  137.424 ± 1.894  ns/op
ContinuableScopeManagerBenchmark.activate        5           0     false  MODERATE  avgt    5  170.942 ± 2.001  ns/op
ContinuableScopeManagerBenchmark.activate        5         100      true     LIGHT  avgt    5  137.123 ± 1.267  ns/op
ContinuableScopeManagerBenchmark.activate        5         100      true  MODERATE  avgt    5  172.344 ± 2.831  ns/op
ContinuableScopeManagerBenchmark.activate        5         100     false     LIGHT  avgt    5  137.976 ± 0.719  ns/op
ContinuableScopeManagerBenchmark.activate        5         100     false  MODERATE  avgt    5  173.601 ± 1.886  ns/op
```

branch `2ea761e86fd5e4eb0d072be0e725af08737b0bca `

```
Benchmark                                  (depth)  (maxDepth)  (strict)    (work)  Mode  Cnt    Score    Error  Units
ContinuableScopeManagerBenchmark.activate        1           0      true     LIGHT  avgt    5   56.458 ±  0.565  ns/op
ContinuableScopeManagerBenchmark.activate        1           0      true  MODERATE  avgt    5   61.324 ±  1.402  ns/op
ContinuableScopeManagerBenchmark.activate        1           0     false     LIGHT  avgt    5   56.157 ±  0.550  ns/op
ContinuableScopeManagerBenchmark.activate        1           0     false  MODERATE  avgt    5   61.243 ±  0.909  ns/op
ContinuableScopeManagerBenchmark.activate        1         100      true     LIGHT  avgt    5   56.169 ±  0.692  ns/op
ContinuableScopeManagerBenchmark.activate        1         100      true  MODERATE  avgt    5   60.926 ±  1.009  ns/op
ContinuableScopeManagerBenchmark.activate        1         100     false     LIGHT  avgt    5   55.888 ±  0.688  ns/op
ContinuableScopeManagerBenchmark.activate        1         100     false  MODERATE  avgt    5   60.926 ±  1.132  ns/op
ContinuableScopeManagerBenchmark.activate        3           0      true     LIGHT  avgt    5   93.311 ±  1.070  ns/op
ContinuableScopeManagerBenchmark.activate        3           0      true  MODERATE  avgt    5  117.095 ±  0.920  ns/op
ContinuableScopeManagerBenchmark.activate        3           0     false     LIGHT  avgt    5   93.416 ±  0.769  ns/op
ContinuableScopeManagerBenchmark.activate        3           0     false  MODERATE  avgt    5  115.838 ±  1.031  ns/op
ContinuableScopeManagerBenchmark.activate        3         100      true     LIGHT  avgt    5   96.108 ±  1.448  ns/op
ContinuableScopeManagerBenchmark.activate        3         100      true  MODERATE  avgt    5  114.675 ±  0.981  ns/op
ContinuableScopeManagerBenchmark.activate        3         100     false     LIGHT  avgt    5   95.840 ±  0.551  ns/op
ContinuableScopeManagerBenchmark.activate        3         100     false  MODERATE  avgt    5  120.067 ±  1.704  ns/op
ContinuableScopeManagerBenchmark.activate        5           0      true     LIGHT  avgt    5  129.881 ±  0.980  ns/op
ContinuableScopeManagerBenchmark.activate        5           0      true  MODERATE  avgt    5  172.945 ± 22.347  ns/op
ContinuableScopeManagerBenchmark.activate        5           0     false     LIGHT  avgt    5  131.886 ±  3.025  ns/op
ContinuableScopeManagerBenchmark.activate        5           0     false  MODERATE  avgt    5  168.731 ±  1.860  ns/op
ContinuableScopeManagerBenchmark.activate        5         100      true     LIGHT  avgt    5  131.646 ±  1.457  ns/op
ContinuableScopeManagerBenchmark.activate        5         100      true  MODERATE  avgt    5  168.686 ±  3.143  ns/op
ContinuableScopeManagerBenchmark.activate        5         100     false     LIGHT  avgt    5  131.964 ±  1.052  ns/op
ContinuableScopeManagerBenchmark.activate        5         100     false  MODERATE  avgt    5  168.375 ±  0.615  ns/op
```

### CODE REVIEW

File: dd-trace-core/src/main/java/datadog/trace/core/scopemanager/ContinuableScopeManager.java [hunk 1]
```diff
@@ -70,26 +70,38 @@ public AgentScope activate(final AgentSpan span, final ScopeSource source) {
...
-    return scope;
+      final ContinuableScope active,
+      final Continuation continuation,
+      final AgentSpan span,
+      final ScopeSource source) {
+    final Scope scope =
+        delegate instanceof ContinuableScopeManager
+            ? ((ContinuableScopeManager) delegate)
+                .handleSpan(active, null, span, ScopeSource.INSTRUMENTATION)
+            : delegate.handleSpan(span);
```
Comment (randomanderson) [inline dd-trace-core/src/main/java/datadog/trace/core/scopemanager/ContinuableScopeManager.java:99]: delegate is never an instance of ContinualScopeManager

**Comment by richardstartin** [comment]
@randomanderson these changes aim to undo a 10% regression since 0.57.0 which became apparent in the reliability environment as soon as the change increasing the size of `ContinuableScopeManager.activate` was merged. It's difficult to provoke in a microbenchmark, and yes, it's very brittle. I'm not sure this adds much complexity really.

**Comment by richardstartin** [comment]
@randomanderson please check if #1678 is more agreeable, it should have the same effect.

**Comment by richardstartin** [comment]
Closing for now

**Pull request closed without merging**